contributors/design-proposals/node/pod-pid-namespace.md

```diff
@@ -1,77 +1,150 @@
+*   Per-pod configuration of PID namespace sharing
+*   Ability to change default sharing behavior in `v2.Pod`
 
-We will modify the Docker implementation of the CRI to use a shared PID
-namespace when running with a version of Docker >= 1.12. The legacy
-`dockertools` implementation will not be changed.
-
-Linking this change to the CRI means that Kubernetes users who care to test such
-changes can test the combined changes at once. Users who do not care to test
-such changes will be insulated by Kubernetes not recommending Docker >= 1.12
-until after switching to the CRI.
-
-Other changes that must be made to support this change:
-
-1. Add a test to verify all containers restart if the infra container
-   responsible for the PodSandbox dies. (Note: With Docker 1.12 if the source
-   of the PID namespace dies all containers sharing that namespace are killed
-   as well.)
-2. Modify the Infra container used by the Docker runtime to reap orphaned
-   zombies ([#36853](https://pr.k8s.io/36853)).
-
-## Rollout Plan
+Non-goals include:
 
-SIG Node is planning to switch to the CRI as a default in 1.6, at which point
-users with Docker >= 1.12 will receive a shared PID namespace by default.
-Cluster administrators will be able to disable this behavior by providing a flag
-to the kubelet which will cause the dockershim to revert to previous behavior.
+*   Creating a general purpose init solution
+*   Multiple shared PID namespaces per pod
+*   Per-container configuration of PID namespace sharing
 
-The ability to disable shared PID namespaces is intended as a way to roll back
-to prior behavior in the event of unforeseen problems. It won't be possible to
-configure the behavior per-pod. We believe this is acceptable because:
+### Configurable PID Namespace Sharing
 
-* We have not identified a concrete use case requiring isolated PID namespaces.
-* Making PID namespace configurable requires changing the CRI, which we would
-  like to avoid since there are no use cases.
+We will add first class support for configuring isolated and shared PID
+namespaces for pods by adding a new boolean field `sharedPID` to the pod spec.
+When set to true, all containers in the pod will share a single PID namespace
+when supported by runtime. When not supported by the runtime, for example with
+docker versions less than 1.13.1, the runtime will respond with an error and the
+pod will not start.
 
-In a future release, SIG Node will recommend docker >= 1.12. Unless a compelling
-use case for isolated PID namespaces is discovered, we will remove the ability
-to disable the shared PID namespace in the subsequent release.
+The Container Runtime Interface (CRI) will be updated to explicitly support
+shared PID namespaces. When `sharedPID` is true, runtimes MUST share a PID
+namespace for the pod or return an error. When `sharedPID` is false, runtimes
+MAY implement isolated PID namespaces. That is, omitting this option defaults to
```

Comment:
Wondering why this is left as a MAY. Can't it be MUST?

Reply:
MAY is the status quo of no explicit default. I left it MAY for two reasons: 

1. Do isolated PID namespaces exist in all runtimes? What do hypervisor runtimes do? What about rkt? I'm not sure since I only use Docker.
2. Do any current runtimes default to shared PID? Do we care about being backwards compatible with them? If yes, then we can't retroactively define a default behavior for v1.Pod, I think.

Reply:
@yujuhong What do you think about an enum for the CRI? Something like: https://github.com/verb/kubernetes/commit/db76a1bc05d582a56f53cb56df3a65e6b236cd13

Reply:
I've rewritten this section to clearly define the behavior.

Reply:
> Do isolated PID namespaces exist in all runtimes? What do hypervisor runtimes do? What about rkt? I'm not sure since I only use Docker.

Hypervisor-based runtime cannot support host PID namespace, etc. We still need more changes even in Kubernetes to fully integrate with them. AFAIK, hypervisor-based runtimes that support docker container images AND Linux should be able to support shared and individual PID namespaces.

/cc @feiskyer 

> Do any current runtimes default to shared PID? Do we care about being backwards compatible with them? If yes, then we can't retroactively define a default behavior for v1.Pod, I think.

This change in the proposal is to ensure the behavior of v1.Pod in Kubernetes does not change, so I think it makes sense to enforce this in kubelet through CRI.

Reply:
> @yujuhong What do you think about an enum for the CRI? Something like: verb/kubernetes@db76a1b

Personally I like that better than a set of booleans for toggling the behavior :-) We could name the `enum` NamespaceType to be more generic.
That is, however, an API breaking change for CRI. 

@mrunalp @Random-Liu @feiskyer thoughts?

Reply:
Ok, great, a default of isolated it is. Good point on the generic name.

Also I realized that we'll need to be able to target a container's namespace for debug containers, which might influence the API design. This could be done as a fourth NamespaceType: https://github.com/verb/kubernetes/commit/48fe4c260c6bf0ba5b4f8b71b71da08d12a9e4d2

Reply:
@feiskyer for frakti? @Random-Liu for containerd @mrunalp for cri-o?

Reply:
At sig-node, we talked about this, and the decision is introducing enum. Thanks!

Reply:
@yujuhong Enum does look cleaner to me. And we could easily change to use the new enum in cri-containerd.